File: Advance/6.Write a program to detect a cycle in a linked list/Solution.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};

bool detectCycle(Node* head) {
    Node *slow = head, *fast = head;
    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast)
            return true;
    }
    return false;
}

int main() {
    int n, val, pos;
    cout << "Enter the number of nodes: ";
    cin >> n;

    Node *head = nullptr, *tail = nullptr;
    for (int i = 0; i < n; i++) {
        cout << "Enter value for node " << i + 1 << ": ";
        cin >> val;
        Node* newNode = new Node(val);
        if (head == nullptr) {
            head = newNode;
        } else {
            tail->next = newNode;
        }
        tail = newNode;
    }

    cout << "Enter the position to create a cycle (0 for no cycle): ";
    cin >> pos;
    if (pos > 0) {
        Node* temp = head;
        for (int i = 1; i < pos; i++)
            temp = temp->next;
        tail->next = temp;
    }

    if (detectCycle(head))
        cout << "Cycle detected in the linked list.\n";
    else
        cout << "No cycle detected in the linked list.\n";

    return 0;
}

File: Advance/6.Write a program to detect a cycle in a linked list/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def detect_cycle(head):
    slow, fast = head, head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

if __name__ == "__main__":
    n = int(input("Enter the number of nodes: "))
    head = None
    tail = None

    for i in range(n):
        val = int(input(f"Enter value for node {i + 1}: "))
        new_node = Node(val)
        if head is None:
            head = new_node
        else:
            tail.next = new_node
        tail = new_node

    pos = int(input("Enter the position to create a cycle (0 for no cycle): "))
    if pos > 0:
        temp = head
        for _ in range(pos - 1):
            temp = temp.next
        tail.next = temp

    if detect_cycle(head):
        print("Cycle detected in the linked list.")
    else:
        print("No cycle detected in the linked list.")
